cv_bridge/src/cv_mat_sensor_msgs_image_type_adapter.cpp

```diff
@@ -202,10 +220,10 @@ ROSCvMatContainer::get_sensor_msgs_msg_image_copy(
         throw std::runtime_error("unsupported encoding type");
     }
   }
-  sensor_msgs_image.step = static_cast<sensor_msgs::msg::Image::_step_type>(frame_.step);
-  size_t size = frame_.step * frame_.rows;
+  sensor_msgs_image.step = static_cast<sensor_msgs::msg::Image::_step_type>(sourceMat.step);
+  size_t size = sourceMat.step * sourceMat.rows;
   sensor_msgs_image.data.resize(size);
-  memcpy(&sensor_msgs_image.data[0], frame_.data, size);
+  memcpy(&sensor_msgs_image.data[0], sourceMat.data, size);
```

Comment:
The only thing here is that we might try to download the sensor_msgs_image directly. We can query the necessary info (from gpu mat and encodings to allocate sensor_msgs image data accordingly such that the data should fit.

Mat ptr and data should be the same memory afterwards.

If the constructors would support also passing in data to share/own (like a sensor-image) with the mat type constructors, you might not even need to allocate data if owned.

It might be good practice to pass in (like for CvImage) shared or unique object ptrs or both that can hold a shared or unique state (i.e., the passed mat is only valid as long as the ptr is valid resp. must dellocate the resource. This way one can safely publish the adapted type and we do not need to copy the mats.

Reply:
Not sure if I fully understand, but I think the issue here is the type. ```cv::Mat``` has an internal store type of ```uchar *``` while the message data is ```vector<uint8_t>``` so the copy is always required (can be in a different form but still needs to exist). Is this what you meant?

Reply:
```cxx
// precompute needed storage capacity
// Default init with 0, but may still be faster than download + copy.
sensor_msg_image.data.resize(precomputedSize);
// or reserve and insert/emplace if faster and implementable with/mat/gpumat

/// Fill other fields of image
/// ...

cv::Mat sourceMat = matFromImage(sensor_msg_image);

///...
gpumat.download(sourceMat);
/// might still copy if download requires a specific layout that does not match the precomputed size.
```